cmd/notafter-backfill/main.go

```diff
@@ -0,0 +1,215 @@
+package main
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"time"
+
+	"github.com/jmhodges/clock"
+	"github.com/letsencrypt/boulder/cmd"
+	"github.com/letsencrypt/boulder/core"
+	blog "github.com/letsencrypt/boulder/log"
+	"github.com/letsencrypt/boulder/metrics"
+	"github.com/letsencrypt/boulder/sa"
+)
+
+type dbAccess interface {
+	Select(holder interface{}, query string, args ...interface{}) ([]interface{}, error)
+	Update(list ...interface{}) (int64, error)
+}
+
+type backfiller struct {
+	dbMap      dbAccess
+	log        blog.Logger
+	clk        clock.Clock
+	dryRun     bool
+	batchSize  uint
+	batchLimit uint
+	sleep      time.Duration
+}
+
+func (b backfiller) printStatus(
+	serial string, notAfter time.Time, cur, total int, start time.Time) {
+	// Should never happen
+	if total <= 0 || cur < 0 || cur > total {
+		b.log.AuditErr(fmt.Sprintf(
+			"invalid cur (%d) or total (%d)\n", cur, total))
+	}
+	completion := (float32(cur) / float32(total)) * 100
+	now := b.clk.Now()
+	elapsed := now.Sub(start)
+	b.log.Info(
+		fmt.Sprintf("Updating %q notAfter to %q. Cert. %d of %d [%.2f%%]. Elapsed: %s",
+			serial, notAfter, cur+1, total, completion, elapsed.String()))
+}
+
+// Finds the email addresses associated with a reg ID
+func (b backfiller) backfill(certStatus core.CertificateStatus) error {
+	updated, err := b.dbMap.Update(&certStatus)
+	if err != nil {
+		return err
+	}
+	if updated != 1 {
+		return fmt.Errorf("backfill Update() did not change exactly 1 row")
+	}
+	return nil
+}
+
+func (b backfiller) findEmpty() ([]core.CertificateStatus, error) {
+	var certs []core.CertificateStatus
+
+	// Note: there is a slight switch-a-roo happening here where the `expires`
+	// field of the `certificates` table is renamed to `notAfter` in order to
+	// populate the portion of a `CertificateStatus` object we wish to update.
+	_, err := b.dbMap.Select(&certs,
+		`SELECT
+			 cs.serial,
+			 cs.subscriberApproved,
+			 cs.status,
+			 cs.ocspLastUpdated,
+			 cs.revokedDate,
+			 cs.revokedReason,
+			 cs.lastExpirationNagSent,
+			 cs.LockCol,
+			 cs.ocspResponse,
+			 certificates.expires AS notAfter
+			 FROM certificates
+			 JOIN certificateStatus AS cs
```

Comment:
I think the JOIN here is going to make things unnecessarily slow. I would instead select from just the certificateStatus table to find rows where notAfter is null, then do a second select by serial to pull the expires from certificates.